cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+	ConfigSource *corev1.NodeConfigSource
+}
+
+type MachineStatus struct {
+	// If the corresponding Node exists, this will point to its object.
+	// +optional
+	NodeRef *api.ObjectReference
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time
+
+	// Indicates whether or not the Machine is fully reconciled. When a
+	// controller observes that the spec has changed and no longer matches
+	// reality, it should update Ready to false before reconciling the
+	// state, and then set back to true when the state matches the spec.
+	Ready bool
+
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a token value suitable for machine interpretation.
+	// +optional
+	ErrorReason *MachineStatusError
+
+	// +optional
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a human readable string to indicate the problem.
+	ErrorMessage *string
+}
+
+type MachineStatusError string
+
+const (
+	// Represents that the combination of configuration in the MachineSpec
+	// is not supported by this cluster. This is not a transient error, but
+	// indicates a state that must be fixed before progress can be made.
+	//
+	// Example: the ProviderConfig specifies an instance type that doesn't exist,
+	InvalidConfigurationMachineError MachineStatusError = "InvalidConfiguration"
+
+	// This indicates that the MachineSpec has been updated in a way that
+	// is not supported for reconciliation on this cluster. The spec may be
+	// completely valid from a configuration standpoint, but the controller
+	// does not support changing the real world state to match the new
+	// spec.
+	//
+	// Example: the responsible controller is not capable of changing the
+	// container runtime from docker to rkt.
+	UnsupportedChangeMachineError MachineStatusError = "UnsupportedChange"
+
+	// This generally refers to exceeding one's quota in a cloud provider,
+	// or running out of physical machines in an on-premise environment.
+	InsufficientResourcesMachineError MachineStatusError = "InsufficientResources"
+
+	// There was an error while trying to create a Node to match this
+	// Machine. This may indicate a transient problem that will be fixed
+	// automatically with time, such as a service outage, or a terminal
+	// error during creation that doesn't match a more specific
+	// MachineStatusError value.
+	//
+	// Example: timeout trying to connect to GCE.
+	CreateMachineError MachineStatusError = "CreateError"
+
+	// An error was encountered while trying to delete the Node that this
+	// Machine represents. This could be a transient or terminal error, but
+	// will only be observable if the provider's Machine controller has
+	// added a finalizer to the object to more gracefully handle deletions.
+	//
+	// Example: cannot resolve EC2 IP address.
+	DeleteMachineError MachineStatusError = "DeleteError"
+)
+
+type MachineVersionInfo struct {
+	// Semantic version of kubelet to run
+	Kubelet string
```

Comment:
Is this an optional value? In most cases you want to match your kubelet and api-server versions.

Reply:
I was thinking that it would be optional at installation time, but that the installer would fill it in with the value that was actually used. Then, tooling built on top of this API can inspect and potentially modify a concrete value here, instead of having to reason about what an empty value here means. I'll document the expectations here more clearly.

Reply:
Can we make this a struct?  `kubelet` to me is an object, not a string.  Is a Struct with a single string a good start?  For instance, we have multiple components.  Also how does this relate not relate to component config.

Reply:
I'm not against making this a struct, but I'm wondering how you see that evolving in the future. What other fields would you envision in the struct? The way this is laid out currently is:

    machine:
      spec:
        versions:
          kubelet: 1.8.0

Are you hoping to have different ways of specifying the version of kubelet to use beyond a single semver, or were hoping to gather together any configuration related to the kubelet into a single struct, rather than having the version be stand-alone here?